File: src/factory.ts
import { Config } from './config';
import { DatabaseAdapter } from './adapters/base';

// 适配器缓存（单例模式）
const adapterCache: Map<string, DatabaseAdapter> = new Map();

export function getAdapter(dbType: string): DatabaseAdapter {
  const type = dbType.toLowerCase();
  
  // 检查缓存
  if (adapterCache.has(type)) {
    return adapterCache.get(type)!;
  }

  // 创建新适配器
  const adapter = createAdapter(type);
  adapterCache.set(type, adapter);
  return adapter;
}

function createAdapter(dbType: string): DatabaseAdapter {
  switch (dbType) {
    case 'mysql': {
      const config = Config.getMysqlConfig();
      if (!config) throw new Error('MYSQL_URL 环境变量未配置');
      const { MySQLAdapter } = require('./adapters/mysql');
      return new MySQLAdapter(config);
    }

    case 'postgres':
    case 'postgresql': {
      const config = Config.getPostgresConfig();
      if (!config) throw new Error('POSTGRES_URL 环境变量未配置');
      const { PostgresAdapter } = require('./adapters/postgres');
      return new PostgresAdapter(config);
    }

    case 'mongodb':
    case 'mongo': {
      const config = Config.getMongodbConfig();
      if (!config) throw new Error('MONGODB_URL 环境变量未配置');
      const { MongoDBAdapter } = require('./adapters/mongodb');
      return new MongoDBAdapter(config);
    }

    case 'redis': {
      const config = Config.getRedisConfig();
      if (!config) throw new Error('REDIS_URL 环境变量未配置');
      const { RedisAdapter } = require('./adapters/redis');
      return new RedisAdapter(config);
    }

    case 'oracle': {
      const config = Config.getOracleConfig();
      if (!config) throw new Error('ORACLE_URL 环境变量未配置');
      const { OracleAdapter } = require('./adapters/oracle');
      return new OracleAdapter(config);
    }

    case 'sqlite': {
      const config = Config.getSqliteConfig();
      if (!config) throw new Error('SQLITE_PATH 环境变量未配置');
      const { SQLiteAdapter } = require('./adapters/sqlite');
      return new SQLiteAdapter(config);
    }

    default:
      throw new Error(`不支持的数据库类型: ${dbType}`);
  }
}

export function listConfiguredDatabases(): Record<string, boolean> {
  return {
    mysql: Config.getMysqlConfig() !== null,
    postgres: Config.getPostgresConfig() !== null,
    mongodb: Config.getMongodbConfig() !== null,
    redis: Config.getRedisConfig() !== null,
    oracle: Config.getOracleConfig() !== null,
    sqlite: Config.getSqliteConfig() !== null
  };
}

export async function closeAllAdapters(): Promise<void> {
  for (const adapter of adapterCache.values()) {
    try {
      await adapter.close();
    } catch {
      // ignore
    }
  }
  adapterCache.clear();
}

File: src/mcp_db/factory.py
"""数据库适配器工厂 - 单例模式管理连接"""

from typing import Dict, Optional
from .config import Config
from .adapters.base import DatabaseAdapter


class AdapterFactory:
    """适配器工厂 - 单例模式复用连接"""
    
    _instances: Dict[str, DatabaseAdapter] = {}
    
    @classmethod
    def get_adapter(cls, db_type: str) -> DatabaseAdapter:
        """
        获取数据库适配器实例（单例）
        
        Args:
            db_type: 数据库类型 (postgres/mysql/mongodb/redis/oracle)
            
        Returns:
            对应的数据库适配器实例
        """
        db_type = db_type.lower()
        
        # 如果已有实例，直接返回
        if db_type in cls._instances:
            return cls._instances[db_type]
        
        # 创建新实例
        adapter = cls._create_adapter(db_type)
        cls._instances[db_type] = adapter
        return adapter
    
    @classmethod
    def _create_adapter(cls, db_type: str) -> DatabaseAdapter:
        """创建适配器实例（延迟导入驱动）"""
        if db_type == "postgres" or db_type == "postgresql":
            config = Config.get_postgres_config()
            if not config:
                raise ValueError("POSTGRES_URL 环境变量未配置")
            from .adapters.postgres import PostgresAdapter
            return PostgresAdapter(config)
        
        elif db_type == "mysql":
            config = Config.get_mysql_config()
            if not config:
                raise ValueError("MYSQL_URL 环境变量未配置")
            from .adapters.mysql import MySQLAdapter
            return MySQLAdapter(config)
        
        elif db_type == "mongodb" or db_type == "mongo":
            config = Config.get_mongodb_config()
            if not config:
                raise ValueError("MONGODB_URL 环境变量未配置")
            from .adapters.mongodb import MongoDBAdapter
            return MongoDBAdapter(config)
        
        elif db_type == "redis":
            config = Config.get_redis_config()
            if not config:
                raise ValueError("REDIS_URL 环境变量未配置")
            from .adapters.redis import RedisAdapter
            return RedisAdapter(config)
        
        elif db_type == "oracle":
            config = Config.get_oracle_config()
            if not config:
                raise ValueError("ORACLE_URL 环境变量未配置")
            from .adapters.oracle import OracleAdapter
            return OracleAdapter(config)
        
        else:
            raise ValueError(f"不支持的数据库类型: {db_type}")
    
    @classmethod
    def list_configured_databases(cls) -> Dict[str, bool]:
        """列出所有已配置的数据库"""
        return {
            "postgres": Config.get_postgres_config() is not None,
            "mysql": Config.get_mysql_config() is not None,
            "mongodb": Config.get_mongodb_config() is not None,
            "redis": Config.get_redis_config() is not None,
            "oracle": Config.get_oracle_config() is not None,
        }
    
    @classmethod
    def close_all(cls) -> None:
        """关闭所有连接"""
        for adapter in cls._instances.values():
            try:
                adapter.close()
            except Exception:
                pass
        cls._instances.clear()


# 便捷函数
def get_adapter(db_type: str) -> DatabaseAdapter:
    """获取数据库适配器"""
    return AdapterFactory.get_adapter(db_type)
